src/node_api.cc

```diff
@@ -68,12 +108,9 @@ class BufferFinalizer : private Finalizer {
       v8::HandleScope handle_scope(finalizer->_env->isolate);
       v8::Context::Scope context_scope(finalizer->_env->context());
```

Comment:
I think the proper fix would be to use a `CallbackScope` here, and letting that take over the error handling. Generally speaking, there should always be a `CallbackScope` present when entering JS code (or other code that may throw JS exceptions) asynchronously.

For `AfterThreadPoolWork`, there already is one – if that doesn’t ensure that uncaught exceptions from inside it are treated as such, then that’s a bug with `CallbackScope`.

Reply:
CallbackScope requires an async resource to trigger async hooks. Whilst this is a finalizer but not an actual "async resource", I don't find it is appropriate for finalizer callback to trigger async hooks. Like, in the callback of FinalizationRegistry, the `async_hooks.executionAsyncId` is always 0. OTOH, we already have `napi_env__.CallIntoModule` to handle the exception for calling into modules (which may call into JavaScript) here.

For `AfterThreadPoolWork`, the change is mainly for alignment across the node-api implementation. We use `CallbackIntoModule` for all async callbacks, `CallFinalizer` for finalizers, and `CallIntoModule` for all other types of calling into modules. The CallbackScope there still counts since it triggers async hooks, while `CallbackIntoModule` doesn't.

Reply:
> Whilst this is a finalizer but not an actual "async resource", I don't find it is appropriate for finalizer callback to trigger async hooks.

Okay, but we’re calling functions asynchronously here – and as you point out, we’re currently losing async context when doing so.

I think it was a bad idea to allow finalizers to call JS in the first place, but we’re stuck with Node-API allowing this, so we should also do it properly.

Reply:
The problem for the finalizers to facilitate CallbackScopes is that they don't have an async id or async resource that async_hooks depends on. (the object itself can not be used as the async resource since on the finalization callback the object is already GC-ed)

This specific site may work as expected as the status-quo. However, the problem been fixed in this PR is that we have to _unify_ the invocation of async callback sites across the node-api implementation _while not breaking existing add-ons_ until the next major release. So I'm just duplicating the try-catch work which has been done in the CallbackScope in the node-api impl again, to enable the finer control with a cli flag.

Reply:
(Also, I'm not sure about thread-safe functions, but for finalizers I’d have expected that it’s currently the user’s job to create a `CallbackScope` or use `MakeCallback`, since Node-API does provide access to these and it’s the right way to access JS in async contexts. It’s silly that that’s necessary, of course, because Node-API should just have done this from the beginning under the hood.)

Reply:
@legendecas Alright, so … big picture, I’m worried that this PR will make it harder to get to away of handling async callbacks in Node-API that is consistent with the rest of the async handling code in Node.js, rather than easier. I’ve been regretting that I didn’t speak up more about this type of issue earlier, because now Node-API has gotten really **bad** when it comes to being consistent about async callback handling – as you point out yourself, we have primitives in the API that do implicit async context tracking, and others which just don’t.

I’ll remove my objection here, but I’d really appreciate if it the Node-API team took some time to take a step back, and figure out what the consistent end state is that you’ll be working towards, and how to get there, rather than only doing incremental fixes like this PR.